components/browser/thumbnails/src/main/java/mozilla/components/browser/thumbnails/utils/ThumbnailDiskCache.kt

```diff
@@ -0,0 +1,72 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.thumbnails.utils
+
+import android.content.Context
+import android.graphics.Bitmap
+import com.jakewharton.disklrucache.DiskLruCache
+import mozilla.components.support.base.log.logger.Logger
+import java.io.File
+import java.io.IOException
+
+private const val MAXIMUM_CACHE_THUMBNAIL_DATA_BYTES: Long = 1024 * 1024 * 100 // 100 MB
+private const val THUMBNAIL_DISK_CACHE_VERSION = 1
+private const val WEBP_QUALITY = 90
+
+/**
+ * Caching bitmaps and resource URLs on disk.
+ */
+class ThumbnailDiskCache {
```

Comment:
I am surprised that we are using the session id as a key here. This means this is not something we could use for top sites etc. too. Fennec had a cache using the URL as a key, I think.

If this is purely for the current list of sessions (e.g. tabs tray) then I wonder if we should integrate this into `SessionStorage` instead of having this as a standalone cache 🤔

Reply:
I suggested using the session ID for the tabs tray use case where we would want a thumbnail per session. I wasn't aware of top sites or other features that would want thumbnails too, but we could look into making sure we support both.

> I wonder if we should integrate this into SessionStorage instead of having this as a standalone cache 🤔

SessionStorage sounds like a good idea. We would need to make more `save` calls whenever a new site loads in order to keep the thumbnail fresh, but maybe that could be abstracted away from the in-memory/disk cache? We haven't reached https://github.com/mozilla-mobile/android-components/issues/6925 and https://github.com/mozilla-mobile/android-components/issues/6905 yet.

Reply:
I am wondering if I should switch to using sessionIdOrUrl as the key value to the disk cache to resolve the different use case.

Reply:
> I am wondering if I should switch to using sessionIdOrUrl as the key value to the disk cache to resolve the different use case.

That sounds like a good idea. Thinking about it more after our conversations, it's probably best to not have the Thumbnails in session storage when we want to re-use this (eventually) for top sites. It would be more difficult to control an in-memory cache along with the session storage as this gets more complicated.